kudu/src/main/java/org/apache/apex/malhar/kudu/scanner/AbstractKuduPartitionScanner.java

```diff
@@ -0,0 +1,240 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractKuduPartitionScanner.class);
+
+  /***
+   * Scans all of the scans planned for the current input operator. Note that the current implementations are
+   * threaded implementations. Depending on the scanner, there are multiple threads or single thread
+   * @param parsedQuery Instance of the parser that parsed the incoming query.
+   * @param setters Setters that are resolved for the POJO class that represents the Kudu row
+   * @return The total number of records that have been scanned for the passed query
+   * @throws IOException
+   */
+  public abstract int scanAllRecords(SQLToKuduPredicatesTranslator parsedQuery,
+      Map<String,Object> setters) throws IOException;
+
+  public AbstractKuduPartitionScanner(AbstractKuduInputOperator<T,C> parentOperatorRunningThisScanner)
+  {
+    parentOperator = parentOperatorRunningThisScanner;
+    apexKuduConnectionInfo = parentOperator.getApexKuduConnectionInfo();
+  }
+
+  /***
+   * Implements common initialization logic across all types of scanners. Currently it is 1. Creating the thread pool
+   *  service for executors and 2. Creating the requisite connection pool for the scanners to use.
+   */
+  public void initScannerCommons()
+  {
+    kuduConsumerExecutor = Executors.newFixedThreadPool(threadPoolExecutorSize);
+    List<KuduPartitionScanAssignmentMeta> allPartitionsThatNeedScan = parentOperator.getPartitionPieAssignment();
+    Collections.sort(allPartitionsThatNeedScan,
+        new Comparator<KuduPartitionScanAssignmentMeta>()
+      {
+        @Override
+        public int compare(KuduPartitionScanAssignmentMeta left, KuduPartitionScanAssignmentMeta right)
+        {
+          return left.getOrdinal() - right.getOrdinal();
+        }
+      });
+    for ( int i = 0; i < threadPoolExecutorSize; i++) {
+      connectionPoolForThreads.put(i,apexKuduConnectionInfo.build());
+    }
+    LOG.info("Scanner running with " + connectionPoolForThreads.size() + " kudu connections");
+  }
+
+  /***
+   * Used to renew a connection in case it is dead. There can be scenarios when there is a continuous sequence of
+   * queries that do not touch a tablet resulting in inactivity on the kudu client session.
+   * @param indexPos The index position in the connection pool
+   * @return A renewed connection in case it was dead
+   */
+  public ApexKuduConnection verifyConnectionStaleness(int indexPos)
+  {
+    ApexKuduConnection apexKuduConnection = connectionPoolForThreads.get(indexPos);
+    checkNotNull(apexKuduConnection, "Null connection not expected while checking staleness of" +
+        " existing connection");
+    if (apexKuduConnection.getKuduSession().isClosed()) {
+      try {
+        apexKuduConnection.close(); // closes the wrapper
+      } catch (Exception e) {
+        LOG.error(" Could not close a possibly stale kudu connection handle ", e);
```

Comment:
if it is an error, then why not throw exception?

Reply:
We do not want to throw an error as this "error" state can be reached if there is no valid query that lands on this Apex partition for a significant amount of time and the client handle socket times out/becomes stale. Hence only logging the stale state and reinitialising the connection using the initially provided connect info object.

Reply:
OK, when this isn't an error then why log it as error? If it is a condition the user cannot do anything about and is of no consequence, why not debug or info?

Reply:
There are two aspects to these few lines of code. The context is we are trying to close a session that is stale and then building a new connection. While invoking the close() method on the connection, it is forcing a checked exception ( and my understanding is that it should not result in an exception even if the session has become stale). Hence was logging it as an error for the aspect of closing the connection ( and only if we get an exception while closing ). The next line is a log statement that that is at level Info stating that the connection is being rebuilt.